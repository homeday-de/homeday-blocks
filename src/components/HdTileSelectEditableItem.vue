<template>
  <hd-tile-select-item
    v-if='!isEditing'
    class="tile-select-editable-item"
    :class="{
      'tile-select-editable-item--selected': isSelected,
    }"
    v-bind="$attrs"
    :value="value"
    @input="enableEditing"
  />
  <input
    v-else
    class="tile-select-editable-item tile-select-editable-item--input"
    :class="{
      'tile-select-editable-item--selected': isSelected,
    }"
    ref="input"
    :value="value"
    v-bind="$attrs"
    @input="$emit('input', $event.target.value)"
    @focus="enableEditing"
    @blur="disableEditing"
    @keydown.enter.prevent.stop="disableEditing"
  >
</template>

<script>
import HdTileSelectItem from 'homeday-blocks/src/components/HdTileSelectItem.vue';

export default {
  name: 'HdTileSelectEditableItem',
  inheritAttrs: false,
  components: {
    HdTileSelectItem,
  },
  props: {
    value: {
      type: [String, Number],
      required: true,
    },
  },
  data() {
    return {
      isEditing: false,
    };
  },
  computed: {
    isSelected() {
      return this.$attrs.model === this.value;
    },
  },
  methods: {
    enableEditing() {
      this.isEditing = true;

      this.$nextTick(() => {
        this.$refs.input.focus();
      });
    },
    disableEditing() {
      this.isEditing = false;
    },
  },
};
</script>

<style lang="scss" scoped>
@import "@/styles/mixins.scss";

.tile-select-editable-item {
  background-color: getShade($quaternary-color, 40);

  &--selected {
    background-color: getShade($secondary-color, 110);
    color: white;
  }

  &--input{
    height: 100%;
    width: 100%;
    border: 0;
    padding: 0;
    font-family: inherit;
    font-size: inherit;
    text-align: center;

    // Hide the browser's controls
    -moz-appearance: textfield;
    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button {
      -webkit-appearance: none;
      margin: 0;
    }
  }
}
</style>
