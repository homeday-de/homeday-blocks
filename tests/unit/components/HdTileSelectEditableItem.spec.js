import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdTileSelectEditableItem from '@/components/HdTileSelectEditableItem.vue';

const model = 1;
const value = 2;
const wrapperFactory = wrapperFactoryBuilder(HdTileSelectEditableItem, {
  props: {
    model,
    value,
  },
});

describe('HdTileSelectEditableItem', () => {
  it('renders component not editable', () => {
    const wrapper = wrapperFactory();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.get('button')).toBeTruthy();
  });

  it('render component editable', () => {
    const wrapper = wrapperFactory({
      data: {
        isEditing: true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.get('input')).toBeTruthy();
  });

  it('emits value on input', () => {
    const newValue = '3';
    const wrapper = wrapperFactory({
      data: {
        isEditing: true,
      },
    });
    wrapper.setValue(newValue);

    expect(wrapper.emitted().input[0][0]).toBe(newValue);
  });
});
