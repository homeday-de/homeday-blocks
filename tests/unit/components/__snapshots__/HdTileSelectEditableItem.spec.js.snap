// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdTileSelectEditableItem render component editable 1`] = `<input model="1" class="tile-select-editable-item tile-select-editable-item--input">`;

exports[`HdTileSelectEditableItem renders component not editable 1`] = `
<button type="button" class="tile-select-item tile-select-editable-item">
  2
</button>
`;
